search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,498 @@
+                  gpointer user_data)
+{
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+    GVariantBuilder meta, metas;
+    gchar *gicon_str;
+    gint i;
+
+    g_variant_builder_init (&metas, G_VARIANT_TYPE ("aa{sv}"));
+
+    for (i = 0; page_ids[i] != NULL; i++){
+      PageData *data = g_hash_table_lookup (app->page_data_hash_map, page_ids[i]);
+      g_variant_builder_init (&meta, G_VARIANT_TYPE ("a{sv}"));
+      g_variant_builder_add (&meta, "{sv}",
+                             "id", g_variant_new_string (page_ids[i]));
+      g_variant_builder_add (&meta, "{sv}",
+                             "name", g_variant_new_string (data->title));
+      gicon_str = g_icon_to_string (data->icon);
+      g_variant_builder_add (&meta, "{sv}",
+                             "gicon", g_variant_new_string (gicon_str));
+      g_free (gicon_str);
+      g_variant_builder_add (&meta, "{sv}",
+                             "description", g_variant_new_string (data->desc));
+      g_variant_builder_add_value (&metas, g_variant_builder_end (&meta));
+    }
+
+    return g_variant_new ("(aa{sv})", &metas);
+}
+
+
+static gboolean
+match_word (const gchar *search_phrase,
+            const gchar *title_casefold,
+            const gchar *desc_casefold)
+{
+    static GRegex *nonword = NULL;
+    gint stri;
+    gchar **strs;
+    gboolean ret = TRUE;
+
+    if (nonword == NULL)
+        nonword = g_regex_new ("\\W", 0, 0, NULL);
+    g_assert (nonword != NULL);
+
+    strs = g_regex_split (nonword, search_phrase, 0);
+    for (stri = 0; strs[stri]; stri++) {
+        if (!title_casefold || !strstr (title_casefold, strs[stri])) {
+            if (!desc_casefold || !strstr (desc_casefold, strs[stri])) {
+                ret = FALSE;
+                break;
+            }
+        }
+    }
+
+    g_strfreev (strs);
+
+    return ret;
+}
+
+static GVariantBuilder
+get_search_results (gchar* term,
+                    gpointer user_data)
+{
+
+    GVariantBuilder builder;
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+    GHashTableIter iter;
+    GPtrArray *res = g_ptr_array_new();
+    gint i;
+    const gchar *page_id;
+    PageData *data;
+
+    g_hash_table_iter_init (&iter, app->page_data_hash_map);
+    g_variant_builder_init (&builder, G_VARIANT_TYPE ("as"));
+
+    while (g_hash_table_iter_next (&iter, (gpointer *)&page_id, (gpointer *)&data)) {
+        if (match_word (term, data->title_casefold, data->desc_casefold))
+            g_variant_builder_add (&builder, "s", page_id);
+    }
+    return builder;
+}
+
+static void
+handle_get_initial_result_set (YelpShellSearchProvider2 *skeleton,
+                               GDBusMethodInvocation *invocation,
+                               gchar **terms,
+                               gpointer user_data)
+{
+    gchar *joined_terms = g_strjoinv (" ", terms);
+
+    if (g_strv_length (terms) < 2 && g_utf8_strlen (terms[0], -1) < 3 ){
```

Comment:
Have seen this other search providers, where they avoid searching 1 or 2 or 3 char. But yes its up to them, I am fine with removing it.